### PR TITLE
casemapping: Move CaseMappingData over to proper ULE type

### PR DESCRIPTION
**Pull request by Manishearth**
First step in casemapping work (Part of https://github.com/unicode-org/icu4x/issues/3481)

Haven't really tested this out but we don't have testcases anyway.

### CODE REVIEW

File: experimental/casemapping/src/provider/data.rs [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+    }
+}
+
+impl AsULE for CaseMappingData {
+    type ULE = CaseMappingDataULE;
+
+    fn from_unaligned(ule: Self::ULE) -> Self {
+        let sixteen = ule.0.as_unsigned_int();
+
+        let ignoreable = (sixteen & CaseMappingDataULE::CASE_IGNOREABLE_BIT) != 0;
```
Comment (sffc) [inline experimental/casemapping/src/provider/data.rs:344]: Question: previously, all of these values were lazily computed. Now you're essentially pre-computing them. Does this matter? (Are there cases where you compute more things than you need?)

**Review by sffc — APPROVED** [review]

**Review by Manishearth — COMMENTED** [review]
oops forgot to submit this comment

File: experimental/casemapping/src/provider/data.rs (hunk 1 above)
Comment (Manishearth) [reply experimental/casemapping/src/provider/data.rs:344]: Potentially. Overall I don't think it matters all that much; and we can optimize with `get_ule()` + specific methods in the future if we need for code that is in tight loops.

The main difference is that this essentially takes a lot of checks that used to be `debug_assert!()`s and sticks them in the type system, which means that check always happens but is cheap.

File: experimental/casemapping/src/provider/data.rs [hunk 2]
```diff
@@ -67,11 +70,250 @@ impl DotType {
...
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct CaseMappingData {
+    pub ignoreable: bool,
+    /// The delta between this code point and its upper/lowercase equivalent.
+    pub kind: CaseMappingDataKind,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum CaseMappingDataKind {
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply experimental/casemapping/src/provider/data.rs:105]: I'm going to merge this but I am very open to do followup renames and such.

**Pull request merged**